Lib/fontTools/designspaceLib/__init__.py

```diff
@@ -434,27 +440,37 @@ def _addRule(self, ruleObject):
         # if none of the conditions have minimum or maximum values, do not add the rule.
         self.rules.append(ruleObject)
         ruleElement  = ET.Element('rule')
-        ruleElement.attrib['name'] = ruleObject.name
-        for cond in ruleObject.conditions:
-            if cond.get('minimum') is None and cond.get('maximum') is None:
-                # neither is defined, don't add this condition
-                continue
-            conditionElement = ET.Element('condition')
-            conditionElement.attrib['name'] = cond.get('name')
-            if cond.get('minimum') is not None:
-                conditionElement.attrib['minimum'] = self.intOrFloat(cond.get('minimum'))
-            if cond.get('maximum') is not None:
-                conditionElement.attrib['maximum'] = self.intOrFloat(cond.get('maximum'))
-            ruleElement.append(conditionElement)
+        if ruleObject.name is not None:
+            ruleElement.attrib['name'] = ruleObject.name
+        for conditions in ruleObject.conditionSets:
+            conditionsetElement = ET.Element('conditionset')
+            for cond in conditions:
+                if cond.get('minimum') is None and cond.get('maximum') is None:
+                    # neither is defined, don't add this condition
+                    continue
+                conditionElement = ET.Element('condition')
+                conditionElement.attrib['name'] = cond.get('name')
+                if cond.get('minimum') is not None:
+                    conditionElement.attrib['minimum'] = self.intOrFloat(cond.get('minimum'))
+                if cond.get('maximum') is not None:
+                    conditionElement.attrib['maximum'] = self.intOrFloat(cond.get('maximum'))
+                conditionsetElement.append(conditionElement)
+            if len(conditionsetElement):
+                ruleElement.append(conditionsetElement)
+        # XXX shouldn't we require at least one sub element?
+        # if not ruleObject.subs:
+        #     raise DesignSpaceDocument('Invalid empty rule with no "sub" elements')
         for sub in ruleObject.subs:
             # skip empty subs
+            # XXX Hm. What does an empty glyphname string even mean?
             if sub[0] == '' and sub[1] == '':
                 continue
             subElement = ET.Element('sub')
             subElement.attrib['name'] = sub[0]
-            subElement.attrib['with'] = sub[1]
+            subElement.attrib['byname'] = sub[1]
```

Comment:
note that changing the attribute name from `with` to `byname` will mean that the latter will appear before the `nama` attribute, since ElementTree sorts the attributes alphabetically when serializing XML. The fonttools xmlWriter allows to keep the order, but has a completely different interface...

Reply:
I'm just not sure when it changed from `with` to `byname`. It can be either.

Reply:
i'm tempted to revert to "with", so that it's sorted logically `<sub name="a" with="a.alt" />`

Reply:
ok